### PR TITLE
Merging client SDK work

### PR DESCRIPTION
**Pull request by angrbrd**
## Description

PR containing the client SDK work.
## Motivation and Context

This pull request contains all of the client sdk work, including all updated unit tests, Makefile updates, and documentation updates.
## How Has This Been Tested?

This pull request introduces or updates three unit test files: registrar.js, chain-test.js, asset-mgmt.js. chain-test.js and asset-mgmt.js may be run in development or network mode for deployment. All tests were automated with the "make node-sdk-unit-tests" command.
## Checklist:
- [x] I have added a [Signed-off-by](https://github.com/hyperledger/fabric/blob/master/CONTRIBUTING.md#legal-stuff)
- [x] Either no new documentation is required by this change, OR I added new documentation
- [x] Either no new tests are required by this change, OR I added new tests
- [x] I have run [goimports](https://godoc.org/golang.org/x/tools/cmd/goimports), [go vet](https://golang.org/cmd/vet/), and [golint](https://github.com/golang/lint). I have cleaned up all valid errors and warnings in code I have added or modified. These tools may generate false positives. Don't be worried about ignoring some errors or warnings. The goal is clean, consistent, and readable code.

Signed-off-by: Anna Derbakova adderbak@us.ibm.com

### CODE REVIEW

File: sdk/node/doc/classes/_crypto_.crypto.html [hunk 1]
```diff
@@ -157,7 +157,7 @@
 						<li class="tsd-description">
 							<aside class="tsd-sources">
 								<ul>
-									<li>Defined in <a href="https://github.com/smithbk/fabric/blob/2d4b64c/sdk/node/src/crypto.ts#L78">crypto.ts:78</a></li>
+									<li>Defined in <a href="https://github.com/angrbrd/fabric/blob/7e4cfbe/sdk/node/src/crypto.ts#L78">crypto.ts:78</a></li>
```
Comment (srderson) [inline sdk/node/doc/classes/_crypto_.crypto.html]: Why are there lots of links to your personal repository in these doc files?

**Comment by angrbrd** [comment]
That's a good question... These are auto-generated files and I'm not 100% sure how that ends up in there. But I see your point. I'll look into into this and see what can be done.

File: sdk/node/doc/classes/_hlc_.eventdeploycomplete.html [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+				<div class="table-cell" id="tsd-search" data-index="../assets/js/search.js" data-base="..">
+					<div class="field">
+						<label for="tsd-search-field" class="tsd-widget search no-caption">Search</label>
+						<input id="tsd-search-field" type="text" />
+					</div>
+					<ul class="results">
+						<li class="state loading">Preparing search index...</li>
+						<li class="state failure">The search index is not available</li>
+					</ul>
+					<a href="../index.html" class="title">Hyperledger OpenBlockChain</a>
```
Comment (binhn) [inline sdk/node/doc/classes/_hlc_.eventdeploycomplete.html]: we shouldn't use this name `Hyperledger OpenBlockChain`

File: sdk/node/doc/classes/_hlc_.eventdeploysubmitted.html [hunk 3]
```diff
@@ -0,0 +1,265 @@
+<!doctype html>
+<html class="default no-js">
+<head>
+	<meta charset="utf-8">
+	<meta http-equiv="X-UA-Compatible" content="IE=edge">
+	<title>EventDeploySubmitted | Hyperledger OpenBlockChain</title>
+	<meta name="description" content="">
```
Comment (binhn) [inline sdk/node/doc/classes/_hlc_.eventdeploysubmitted.html]: Many instances of `Hyperledger OpenBlockChain`. 
Change it to `Hyperledger Fabric`

**Comment by adecaro** [comment]
Hi @angrbrd, how did you test deploy in network mode? In particular, how did you setup the fabric network? Actually, I'm having trouble in setting up the fabric using vagrant. Nodejs, npm and typings versions are old.

BTW, I have seen that there are some dependencies missing in the file package.json like protobujs and lodash. Am I missing something?

**Comment by smithbk** [comment]
@adecaro The sdk/node/bin/run-unit-tests.sh script runs tests in both network mode and dev mode.  You can look at that script to see how to set up the directory structure for deploy in network mode.  See the prepareExampleForDeployInNetworkMode function.  Maybe we should provide that as a standalone script to make it easier for users to set up other examples for deploy.

**Comment by smithbk** [comment]
@adecaro I should have mentioned that setting up the directory structure is described in the README.md

**Comment by adecaro** [comment]
@smithbk Sorry, I'm confused. _run-unit-tests.sh_ is not mentioned in the README file. 

I got how to prepare the folder structure to deploy a chaincode in network mode. That's fine.

Anyway, even before that, when I run _make node-sdk_ in a clean vagrant instance, it just fails with the following message:

```
vagrant@ubuntu-1404:/opt/gopath/src/github.com/hyperledger/fabric$ make node-sdk
cp ./protos/*.proto ./sdk/node/lib/protos
cp ./membersrvc/protos/*.proto ./sdk/node/lib/protos
cd ./sdk/node && npm install && sudo npm install -g typescript && sudo npm install typings --global && typings install
npm WARN package.json hlc@0.0.9 No repository field.
npm WARN package.json hlc@0.0.9 No license field.
npm ERR! Linux 3.13.0-86-generic
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install"
npm ERR! node v0.12.7
npm ERR! npm  v2.11.3
npm ERR! path ../har-validator/bin/har-validator
npm ERR! code EPROTO
npm ERR! errno -71

npm ERR! EPROTO, symlink '../har-validator/bin/har-validator'
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>
npm ERR! Linux 3.13.0-86-generic
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install"
npm ERR! node v0.12.7
npm ERR! npm  v2.11.3
npm ERR! path npm-debug.log.b14eda52099085229a86ec264b001337
npm ERR! code ETXTBSY
npm ERR! errno -26

npm ERR! ETXTBSY, rename 'npm-debug.log.b14eda52099085229a86ec264b001337'
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /opt/gopath/src/github.com/hyperledger/fabric/sdk/node/npm-debug.log
make: *** [node-sdk] Error 185
```

Any idea on what I'm missing?

**Comment by angrbrd** [comment]
@adecaro the run-unit-tests.sh script is what gets triggered by the `make node-sdk-unit-tests` command. That command is described in the README.md. If you take a look at that run-unit-tests.sh script, it will run through chaincode deployments in both network mode as well as development mode.

I am setting up a clean vagrant environment right now based on my branch with these changes. I will touch base shortly.

**Comment by adecaro** [comment]
Thanks @angrbrd :)

**Comment by rameshthoomu** [comment]
@angrbrd: Travis builds are failing with npm command not found error message as the npm is not installed in travis. Updated below npm command and it worked: 
`cd ./sdk/node && sudo apt-get install -y npm && sudo npm install -g typescript && sudo npm install typings --global && typings install`. 

I could see an issue with membersrvc failed to start: Below is the error message:
`Documentation generated at /home/travis/gopath/src/github.com/hyperledger/fabric/sdk/node/doc
ERROR: member services failed to start; see /tmp/node-sdk-unit-test/membersrvc.log`
**Travis Build for reference:**
"https://travis-ci.org/rameshthoomu/fabric/builds/137870434"

**Comment by rameshthoomu** [comment]
Regarding this issue: membersrvc image is not created before run node-sdk-tests. We have create membersrvc image first and then run node-sdk-tests. Reference Makefile: `unit-test: peer-image gotools node-sdk-unit-tests`. We have to update above command to run `make membersrvc` and `peer-membersrvc-image.`

**Comment by angrbrd** [comment]
@adecaro I just pushed another change should hopefully address any issues you are seeing. The things you need to do in a clean vagrant environment to run the unit tests from this branch are below. Please give this a try and let me know if it works for you.

```
cd $GOPATH/src/github.com/hyperledger/fabric/membersrvc
go build
cd $GOPATH/src/github.com/hyperledger/fabric/peer
go build
cd $GOPATH/src/github.com/hyperledger/fabric
make node-sdk-unit-tests
```

**Comment by rjones-lf** [comment]
reverify jenkins

**Comment by rjones-lf** [comment]
reverify jenkins

**Comment by rjones-lf** [comment]
reverify jenkins

**Comment by smithbk** [comment]
CI is still failing here?  Is anyone investigating?

**Comment by rameshthoomu** [comment]
@angrbrd is looking these issues.

**Comment by angrbrd** [comment]
@binhn and I discussed and agreed that we will merge this without Travis running at this time. The issue with Travis is not obvious and all the unit tests are passing locally. I have created another branch that contains identical changes, but removes the Travis CI integration. Therefore, I will now close this PR and open a new one from that new branch.

**Pull request closed without merging**